File: apps/api/src/agents/multi-agent/graph.ts
/**
 * 面试多 Agent 图拓扑定义
 *
 * 基于 LangGraph StateGraph v1.x
 *
 * 拓扑结构：
 *
 *   START
 *     │
 *     ▼
 *   supervisor ──→ planner ──→ executor ──→ replanner ──→ reviewer
 *     │                              │            │            │
 *     └→ respond_directly            │            │            │
 *                                   └────────────┘            │
 *                                   (replan → executor)       │
 *                                                                │
 *                                            reviewer ──→ planner
 *                                            (revise → replan) │
 *                                                                │
 *                                              reviewer → END
 *                                              (approved)
 *
 *   reviewer ──→ hitl_review ──→ END (HR approved)
 *                (interrupt)
 *
 * 条件边：
 * 1. supervisor → planner | respond_directly   （按意图路由）
 * 2. replanner → executor | reviewer           （按执行状态路由）
 * 3. reviewer → END | planner | hitl_review    （通过/打回/HITL中断）
 * 4. hitl_review → END                         （interrupt 暂停，HR 审批后 Command(resume) 恢复）
 */
import { StateGraph, END, START, interrupt, Command } from '@langchain/langgraph';
import { PostgresSaver } from '@langchain/langgraph-checkpoint-postgres';
import { InterviewAgentState, type InterviewAgentStateType } from './state';
export type { InterviewAgentStateType } from './state';
import { createSupervisorNode, supervisorRouter } from './nodes/supervisor';
import { createPlannerNode } from './nodes/planner';
import { createExecutorNode } from './nodes/executor';
import { createReplannerNode, replannerRouter } from './nodes/replanner';
import { createReviewerNode, reviewerRouter } from './nodes/reviewer';
import { LlmGatewayChatModel } from './llm-gateway-chat-model';
import { HumanMessage, AIMessage, BaseMessage } from 'langchain';
import { BaseCheckpointSaver } from '@langchain/langgraph-checkpoint';
export interface StreamChunk {
  type: 'token' | 'step' | 'final_response' | 'error' | 'hitl_pending';
  content?: string;
  step?: string;
  node?: string;
  error?: string;
  hitlData?: { interviewId: string; score: number; issues: string[] };
}

/**
 * 构建面试多 Agent 图
 *
 * @param model - ChatOpenAI 实例（传入已配置好的 Qwen/DeepSeek 模型）
 * @returns 编译好的 LangGraph CompiledStateGraph
 *
 * @example
 * ```typescript
 * const model = new ChatOpenAI({
 *   modelName: 'qwen-plus',
 *   apiKey: process.env.QWEN_API_KEY,
 *   configuration: { baseURL: 'https://dashscope.aliyuncs.com/compatible-mode/v1' },
 * });
 *
 * const graph = buildInterviewGraph(model);
 * const result = await graph.invoke({
 *   messages: [new HumanMessage('帮我面试一下 AI Agent 工程师')],
 * });
 *
 * console.log(result.final_response);
 * ```
 */
export function buildInterviewGraph(
  model: LlmGatewayChatModel,
  checkpointer?: BaseCheckpointSaver,
) {
    // 创建节点函数
    const supervisorNode = createSupervisorNode(model);
    const plannerNode = createPlannerNode(model);
    const executorNode = createExecutorNode(model);
    const replannerNode = createReplannerNode(model);
    const reviewerNode = createReviewerNode(model);

    // respond_directly 节点：general_qa 直接调 LLM 回复，不走规划
    const respondDirectlyNode = async (
        state: InterviewAgentStateType,
        config?: any,
    ): Promise<Partial<InterviewAgentStateType>> => {
        const lastMessage =
            state.messages[state.messages.length - 1]?.content ?? '';
        const response = await model.invoke([
            {
                role: 'system',
                content:
                    '你是一位专业的 AI 面试官小面。请简洁友好地回答用户的问题，自然口语化，不要用 Markdown 标题或列表。',
            },
            { role: 'user', content: lastMessage },
        ], config);
        const text = response.content as string;
        // 既要 push AIMessage 到 messages（让 state 完整 + checkpointer 持久化完整对话），
        // 也要 set final_response（向后兼容 controller 取值）
        return {
            messages: [new AIMessage(text)],
            final_response: text,
        };
    };

    // hitl_review 节点：评分争议时 interrupt 暂停，等待 HR 审批
    // HR 审批后通过 Command(resume) 恢复，resume value = 'approved' | 'rejected'
    const hitlReviewNode = async (
        state: InterviewAgentStateType,
    ): Promise<Partial<InterviewAgentStateType>> => {
        // 调用 interrupt() 暂停图执行，等待外部输入
        // resume value 由 HR 审批端点传入：'approved' 或 'rejected'
        const verdict = (interrupt as any)(
            'HITL: 评分争议，等待 HR 审批',
        ) as 'approved' | 'rejected';

        if (verdict === 'approved') {
            // HR 批准：使用 reviewer 的草稿 final_response
            return {
                messages: [new AIMessage(state.final_response)],
                hitl_pending: false,
                hitl_verdict: 'approved',
                // final_response 保持不变（reviewer 已设置）
            };
        }

        // HR 拒绝：打回重做
        return {
            final_response: '',
            hitl_pending: false,
            hitl_verdict: 'rejected',
        };
    };

    // hitl_review 条件路由：approved → END，rejected → planner
    const hitlReviewRouter = (
        state: InterviewAgentStateType,
    ): 'end' | 'planner' => {
        if (state.hitl_verdict === 'approved' && state.final_response) {
            return 'end';
        }
        return 'planner';
    };

    // 构建 StateGraph
    const graph = new StateGraph(InterviewAgentState)
        // ── 注册节点 ──
        .addNode('supervisor', supervisorNode)
        .addNode('planner', plannerNode)
        .addNode('executor', executorNode)
        .addNode('replanner', replannerNode)
        .addNode('reviewer', reviewerNode)
        .addNode('respond_directly', respondDirectlyNode)
        .addNode('hitl_review', hitlReviewNode)

        // ── 定义边 ──

        // START → supervisor
        .addEdge(START, 'supervisor')

        // supervisor 条件路由
        .addConditionalEdges('supervisor', supervisorRouter, {
            planner: 'planner',
            respond_directly: 'respond_directly',
        })

        // planner → executor（固定边：规划完就执行）
        .addEdge('planner', 'executor')

        // executor → replanner（固定边：执行完就判断下一步）
        .addEdge('executor', 'replanner')

        // replanner 条件路由
        .addConditionalEdges('replanner', replannerRouter, {
            executor: 'executor',
            reviewer: 'reviewer',
        })

        // reviewer 条件路由（含 HITL 中断）
        .addConditionalEdges('reviewer', reviewerRouter, {
            end: END,
            planner: 'planner',
            hitl_review: 'hitl_review',
        })

        // hitl_review 条件路由（HR 审批后恢复）
        .addConditionalEdges('hitl_review', hitlReviewRouter, {
            end: END,
            planner: 'planner',
        })

        // respond_directly → END
        .addEdge('respond_directly', END);

    return graph.compile({ checkpointer });
}

File: apps/py-api/app/agents/graph.py
"""
LangGraph v0.5 多 Agent 拓扑（对齐 NestJS graph.ts）

拓扑：
   START
     ↓
   supervisor ──→ planner ──→ executor ──→ replanner ──→ reviewer
     │                              │            │            │
     └→ respond_directly            │            │            │
                                   └────────────┘            │
                                                                ↓
                                            reviewer ──→ planner (revise)
                                                                ↓
                                              reviewer → END (approved)
                                              reviewer ──→ hitl_review ──→ END
"""
from typing import Literal
from langgraph.graph import StateGraph, START, END
from langgraph.checkpoint.postgres.aio import AsyncPostgresSaver
from langgraph.types import interrupt, Command

from app.agents.state import InterviewAgentState
from app.agents.nodes.supervisor import supervisor_node, supervisor_router
from app.agents.nodes.planner import planner_node
from app.agents.nodes.executor import executor_node
from app.agents.nodes.replanner import replanner_node, replanner_router
from app.agents.nodes.reviewer import reviewer_node, reviewer_router
from app.llm.qwen_provider import QwenProvider
from app.memory.redis_memory import RedisMemory
from app.memory.milvus_memory import MilvusMemory
from app.memory.mem0_memory import Mem0Memory


INTERVIEW_GRAPH_RECURSION_LIMIT = 60


async def build_interview_graph(
    redis_mem: RedisMemory,
    milvus_mem: MilvusMemory,
    mem0_mem: Mem0Memory,
    settings,
    checkpointer: AsyncPostgresSaver = None,
):
    """
    构建面试多 Agent 图（5 节点 + 2 旁路）

    Returns:
        编译好的 CompiledStateGraph，可直接 graph.stream() / graph.invoke()
    """
    # 初始化 LLM Provider（Qwen via dashscope）
    llm = QwenProvider(
        api_key=settings.QWEN_API_KEY,
        base_url=settings.QWEN_BASE_URL,
        model_name="qwen-plus",
    )

    # 注入记忆到各节点的闭包（LangGraph 节点必须是 async）
    async def _supervisor(state, config=None):
        return await supervisor_node(state, llm, config, redis_mem)

    async def _planner(state, config=None):
        return await planner_node(state, llm, config)

    async def _executor(state, config=None):
        return await executor_node(state, llm, config, milvus_mem, mem0_mem)

    # replanner 是纯路由函数，不调 LLM，直接同步返回
    def _replanner(state, config=None):
        return replanner_node(state, config)

    async def _reviewer(state, config=None):
        return await reviewer_node(state, llm, config, redis_mem)

    # respond_directly 旁路：general_qa 直接 LLM 回复
    async def respond_directly_node(state):
        last_msg = state["messages"][-1]
        user_text = last_msg.content if hasattr(last_msg, "content") else str(last_msg)
        response = await llm.chat([
            {"role": "system", "content": "你是 AI 面试官小面，简洁友好回答，不带 Markdown。"},
            {"role": "user", "content": user_text},
        ])
        from langchain_core.messages import AIMessage
        return {
            "messages": [AIMessage(content=response)],
            "final_response": response,
        }

    # hitl_review 旁路：评分争议 interrupt 暂停
    async def hitl_review_node(state):
        verdict = interrupt("HITL: 评分争议，等待 HR 审批")
        if verdict == "approved":
            from langchain_core.messages import AIMessage
            return {
                "messages": [AIMessage(content=state["final_response"])],
                "hitl_pending": False,
                "hitl_verdict": "approved",
            }
        return {
            "final_response": "",
            "hitl_pending": False,
            "hitl_verdict": "rejected",
        }

    def hitl_review_router(state) -> Literal["end", "planner"]:
        if state.get("hitl_verdict") == "approved" and state.get("final_response"):
            return "end"
        return "planner"

    # 构建 StateGraph
    graph = (
        StateGraph(InterviewAgentState)
        # 注册节点
        .add_node("supervisor", _supervisor)
        .add_node("planner", _planner)
        .add_node("executor", _executor)
        .add_node("replanner", _replanner)
        .add_node("reviewer", _reviewer)
        .add_node("respond_directly", respond_directly_node)
        .add_node("hitl_review", hitl_review_node)
        # 边
        .add_edge(START, "supervisor")
        .add_conditional_edges(
            "supervisor", supervisor_router,
            {"planner": "planner", "respond_directly": "respond_directly"},
        )
        .add_edge("planner", "executor")
        .add_edge("executor", "replanner")
        .add_conditional_edges(
            "replanner", replanner_router,
            {"executor": "executor", "reviewer": "reviewer"},
        )
        .add_conditional_edges(
            "reviewer", reviewer_router,
            {"end": END, "planner": "planner", "hitl_review": "hitl_review"},
        )
        .add_conditional_edges(
            "hitl_review", hitl_review_router,
            {"end": END, "planner": "planner"},
        )
        .add_edge("respond_directly", END)
    )

    return graph.compile(checkpointer=checkpointer)